File: src/statusline.js
const { execFileSync } = require('node:child_process');
const fs = require('node:fs');
const path = require('node:path');

function number(...values) {
  for (const value of values) {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return null;
}

function git(args, cwd) {
  try { return execFileSync('git', args, { cwd, encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'] }).trim(); }
  catch { return ''; }
}

function gitStats(cwd) {
  const root = git(['rev-parse', '--show-toplevel'], cwd);
  if (!root) return { project: path.basename(cwd), branch: '', changes: 0, age: '', added: 0, removed: 0 };
  const status = git(['status', '--porcelain'], cwd);
  const lastCommit = git(['log', '-1', '--format=%ct'], cwd);
  const age = lastCommit ? formatAge(Date.now() - Number(lastCommit) * 1000) : '';
  let added = 0, removed = 0;
  for (const line of git(['diff', '--numstat', 'HEAD'], cwd).split(/\r?\n/)) {
    const [a, d] = line.split(/\s+/);
    if (/^\d+$/.test(a)) added += Number(a);
    if (/^\d+$/.test(d)) removed += Number(d);
  }
  return { project: path.basename(root), branch: git(['branch', '--show-current'], cwd), changes: status ? status.split(/\r?\n/).length : 0, age, added, removed };
}

function formatAge(ms) {
  if (!Number.isFinite(ms) || ms < 0) return '';
  const days = Math.floor(ms / 86400000);
  if (days) return `${days}d`;
  const hours = Math.floor(ms / 3600000);
  if (hours) return `${hours}h`;
  return `${Math.max(1, Math.floor(ms / 60000))}m`;
}

function duration(start) {
  const value = Date.parse(start || '') || number(start) * (String(start).length < 13 ? 1000 : 1);
  return value ? formatAge(Date.now() - value) : '';
}

function pct(value) {
  const n = number(value);
  return n === null ? '--' : `${Math.round(n)}%`;
}

function compactTokens(value) {
  const n = number(value);
  if (n === null) return '--';
  if (n >= 1000000) return `${(n / 1000000).toFixed(1)}m`;
  if (n >= 1000) return `${(n / 1000).toFixed(1)}k`;
  return String(Math.round(n));
}

// Claude's signature orange (#D97757), with a 256-color fallback.
const orange = /^(truecolor|24bit)$/i.test(process.env.COLORTERM || '') ? '\x1b[38;2;217;119;87m' : '\x1b[38;5;173m';

const themes = {
  // Orange stays the identity accent (model, project); everything else gets
  // its own hue so the line reads as a small dashboard, not a grey wall.
  default: { reset: '\x1b[0m', gray: '\x1b[90m', cyan: '\x1b[96m', blue: '\x1b[94m', green: '\x1b[92m', yellow: '\x1b[93m', red: '\x1b[91m', magenta: '\x1b[95m', orange },
  ocean: { reset: '\x1b[0m', gray: '\x1b[90m', cyan: '\x1b[96m', blue: '\x1b[94m', green: '\x1b[92m', yellow: '\x1b[93m', red: '\x1b[91m', magenta: '\x1b[95m' },
  dracula: { reset: '\x1b[0m', gray: '\x1b[90m', cyan: '\x1b[96m', blue: '\x1b[95m', green: '\x1b[92m', yellow: '\x1b[93m', red: '\x1b[91m', magenta: '\x1b[95m' },
  nord: { reset: '\x1b[0m', gray: '\x1b[37m', cyan: '\x1b[96m', blue: '\x1b[94m', green: '\x1b[92m', yellow: '\x1b[93m', red: '\x1b[91m', magenta: '\x1b[95m' },
  mono: { reset: '\x1b[0m', gray: '\x1b[90m', cyan: '\x1b[37m', blue: '\x1b[37m', green: '\x1b[37m', yellow: '\x1b[97m', red: '\x1b[97m', magenta: '\x1b[97m' },
};

const ansi = themes[process.env.CLAUDE_COCKPIT_THEME || 'default'] || themes.default;
ansi.orange = ansi.orange || ansi.cyan;

function color(value, code) {
  if (process.env.NO_COLOR || !value) return String(value ?? '');
  return `${code}${value}${ansi.reset}`;
}

function level(value) {
  const n = number(value);
  if (n === null) return ansi.gray;
  if (n >= 80) return ansi.red;
  if (n >= 50) return ansi.yellow;
  return ansi.green;
}

// withDate: the 7-day reset can be days out, so a bare time is ambiguous —
// prefix the date. The 5-hour reset is always same-day, so it skips this.
function resetClock(value, withDate = false) {
  const n = number(value);
  if (n === null) return '';
  const tz = process.env.CLAUDE_COCKPIT_TZ || 'Asia/Bangkok';
  const time = new Intl.DateTimeFormat('en-GB', { timeZone: tz, hour: '2-digit', minute: '2-digit', hour12: false }).format(new Date(n * 1000));
  if (!withDate) return time;
  const date = new Intl.DateTimeFormat('en-GB', { timeZone: tz, month: 'short', day: 'numeric' }).format(new Date(n * 1000));
  return `${date} ${time}`;
}

function formatDurationMs(value) {
  const n = number(value);
  if (n === null || n < 0) return '';
  const totalMinutes = Math.floor(n / 60000);
  return `${String(Math.floor(totalMinutes / 60)).padStart(2, '0')}:${String(totalMinutes % 60).padStart(2, '0')}`;
}

// Visible width (ANSI codes take no columns) and a padder that uses it — lets
// line 2's fields line up under line 1's, column for column.
function visLen(s) { return s.replace(/\x1b\[[0-9;]*m/g, '').length; }
function padTo(s, w) { return s + ' '.repeat(Math.max(0, w - visLen(s))); }
// A long session name or branch shouldn't blow out every column's width —
// cap it before it enters the alignment math.
function truncate(s, max) { return s.length > max ? `${s.slice(0, max - 1)}…` : s; }

function statusline(input = {}) {
  const cwd = input.workspace?.current_dir || input.cwd || process.cwd();
  const gitInfo = gitStats(cwd);
  const model = input.model?.display_name || input.model?.name || input.model || '--';
  const context = input.context_window || input.context || {};
  const cost = input.cost || {};
  const rate = input.rate_limits || input.rateLimits || {};
  const five = rate.five_hour || rate['5h'] || rate.fiveHour || {};
  const seven = rate.seven_day || rate['7d'] || rate.sevenDay || {};
  const tokens = number(input.total_tokens, input.tokens, cost.total_tokens,
    number(context.total_input_tokens) + number(context.total_output_tokens),
    number(input.usage?.input_tokens) + number(input.usage?.output_tokens));
  const session = input.session_start_time || input.session_start_timestamp || input.sessionStartTime;
  const clock = new Intl.DateTimeFormat('en-GB', { timeZone: process.env.CLAUDE_COCKPIT_TZ || 'Asia/Bangkok', hour: '2-digit', minute: '2-digit', hour12: false, timeZoneName: 'short' }).format(new Date());
  const elapsed = formatDurationMs(cost.total_duration_ms) || duration(session);
  const sep = color(' · ', ansi.gray);
  const added = number(cost.total_lines_added, gitInfo.added) || 0;
  const removed = number(cost.total_lines_removed, gitInfo.removed) || 0;
  const fiveReset = resetClock(five.resets_at ?? five.reset_at);
  const sevenReset = resetClock(seven.resets_at ?? seven.reset_at, true);
  const contextPct = context.used_percentage ?? context.usedPercent ?? context.percentage;
  const fivePct = five.used_percentage ?? five.usedPercent ?? five.percentage;
  const sevenPct = seven.used_percentage ?? seven.usedPercent ?? seven.percentage;
  const costText = `$${number(cost.total_cost_usd, cost.totalCostUsd)?.toFixed(2) || '0.00'}`;
  const meter = (label, value, labelColor = ansi.gray) => `${color(label, labelColor)} ${color(pct(value), level(value))}`;

  // The project/session name goes last and unaligned — it's shown in full,
  // however long, instead of forcing every other column to stretch with it.
  const projectField = input.session_name
    ? `${color(gitInfo.project, ansi.orange)}${color(`@${input.session_name}`, ansi.magenta)}`
    : color(gitInfo.project, ansi.orange);
  // Branch and its change count share one column: "main (1±)".
  const branchField = gitInfo.branch
    ? `${color(truncate(gitInfo.branch, 16), ansi.orange)}${gitInfo.changes ? ` ${color(`(${gitInfo.changes}±)`, ansi.yellow)}` : ''}`
    : color('--', ansi.gray);
  const cols1 = [
    branchField,
    `${color(`+${added}`, ansi.green)} ${color(`-${removed}`, ansi.red)}`,
    color(truncate(model, 16), ansi.orange),
    color(`${clock}${elapsed ? ` ${elapsed}` : ''}`, ansi.blue),
    projectField,
  ];
  const cols2 = [
    meter('ctx', contextPct, ansi.cyan),
    color(compactTokens(tokens), ansi.cyan),
    `${meter('5h', fivePct, ansi.blue)}${fiveReset ? ` ${color(fiveReset, ansi.blue)}` : ''}`,
    `${meter('7d', sevenPct, ansi.magenta)}${sevenReset ? ` ${color(sevenReset, ansi.magenta)}` : ''}`,
    color(costText, ansi.green),
  ];
  // Cap each column so one unusually long field can't stretch every other
  // column with it — a field past the cap just overflows on its own row. The
  // two rows can have different field counts (they do now); widths just
  // covers whichever is longer at each index.
  const n = Math.max(cols1.length, cols2.length);
  const widths = Array.from({ length: n },
    (_, i) => Math.min(32, Math.max(visLen(cols1[i] || ''), visLen(cols2[i] || ''))));
  const align = cols => cols.map((c, i) => (i === cols.length - 1 ? c : padTo(c, widths[i])));
  const line1 = align(cols1).join(sep);
  const line2 = align(cols2).join(sep);
  return `${line1}\n${line2}`;
}

function main() {
  let raw = '';
  try { raw = fs.readFileSync(0, 'utf8'); } catch {}
  let input = {};
  try { input = raw.trim() ? JSON.parse(raw) : {}; } catch {}
  process.stdout.write(statusline(input));
}

module.exports = { statusline, gitStats, main };
if (require.main === module) main();

File: rust/src/statusline.rs
// The Claude Code statusline renderer. Claude Code pipes a JSON payload on
// stdin and expects a two-line string back. Ported from src/statusline.js —
// see that file's comments for the design rationale (why fields go where).
use chrono::{Offset, TimeZone};
use serde_json::Value;
use std::process::{Command, Stdio};

fn number(values: &[Option<f64>]) -> Option<f64> {
    values.iter().flatten().find(|v| v.is_finite()).copied()
}

fn git(args: &[&str], cwd: &str) -> String {
    Command::new("git")
        .args(args)
        .current_dir(cwd)
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .output()
        .ok()
        .filter(|o| o.status.success())
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .unwrap_or_default()
}

struct GitInfo {
    project: String,
    branch: String,
    changes: u32,
    added: u32,
    removed: u32,
}

fn git_stats(cwd: &str) -> GitInfo {
    let root = git(&["rev-parse", "--show-toplevel"], cwd);
    if root.is_empty() {
        let project = std::path::Path::new(cwd).file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
        return GitInfo { project, branch: String::new(), changes: 0, added: 0, removed: 0 };
    }
    let status = git(&["status", "--porcelain"], cwd);
    let (mut added, mut removed) = (0u32, 0u32);
    for line in git(&["diff", "--numstat", "HEAD"], cwd).lines() {
        let mut parts = line.split_whitespace();
        if let (Some(a), Some(d)) = (parts.next(), parts.next()) {
            if let Ok(n) = a.parse::<u32>() {
                added += n;
            }
            if let Ok(n) = d.parse::<u32>() {
                removed += n;
            }
        }
    }
    let project = std::path::Path::new(&root).file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
    let changes = if status.is_empty() { 0 } else { status.lines().count() as u32 };
    GitInfo { project, branch: git(&["branch", "--show-current"], cwd), changes, added, removed }
}

fn format_age(ms: f64) -> String {
    if !ms.is_finite() || ms < 0.0 {
        return String::new();
    }
    let days = (ms / 86_400_000.0).floor();
    if days >= 1.0 {
        return format!("{days:.0}d");
    }
    let hours = (ms / 3_600_000.0).floor();
    if hours >= 1.0 {
        return format!("{hours:.0}h");
    }
    format!("{:.0}m", (ms / 60_000.0).floor().max(1.0))
}

fn duration_since(start: Option<&str>) -> String {
    let Some(start) = start.filter(|s| !s.is_empty()) else { return String::new() };
    let ms = chrono::DateTime::parse_from_rfc3339(start).map(|d| d.timestamp_millis() as f64).ok().or_else(|| {
        start.parse::<f64>().ok().map(|n| if start.len() < 13 { n * 1000.0 } else { n })
    });
    match ms {
        Some(ms) => format_age(now_ms() - ms),
        None => String::new(),
    }
}

fn now_ms() -> f64 {
    std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_millis() as f64
}

fn pct(value: Option<f64>) -> String {
    match value {
        Some(n) => format!("{:.0}%", n.round()),
        None => "--".to_string(),
    }
}

fn compact_tokens(value: Option<f64>) -> String {
    match value {
        None => "--".to_string(),
        Some(n) if n >= 1_000_000.0 => format!("{:.1}m", n / 1_000_000.0),
        Some(n) if n >= 1_000.0 => format!("{:.1}k", n / 1_000.0),
        Some(n) => format!("{:.0}", n.round()),
    }
}

struct Ansi {
    reset: &'static str,
    gray: &'static str,
    cyan: &'static str,
    blue: &'static str,
    green: &'static str,
    yellow: &'static str,
    red: &'static str,
    magenta: &'static str,
    orange: &'static str,
}

fn orange_code() -> &'static str {
    let truecolor = matches!(std::env::var("COLORTERM").unwrap_or_default().to_lowercase().as_str(), "truecolor" | "24bit");
    if truecolor { "\x1b[38;2;217;119;87m" } else { "\x1b[38;5;173m" }
}

fn theme() -> Ansi {
    // Orange stays the identity accent (model, project); everything else
    // gets its own hue so the line reads as a small dashboard, not grey.
    match std::env::var("CLAUDE_COCKPIT_THEME").unwrap_or_default().as_str() {
        "ocean" => Ansi { reset: "\x1b[0m", gray: "\x1b[90m", cyan: "\x1b[96m", blue: "\x1b[94m", green: "\x1b[92m", yellow: "\x1b[93m", red: "\x1b[91m", magenta: "\x1b[95m", orange: "\x1b[96m" },
        "dracula" => Ansi { reset: "\x1b[0m", gray: "\x1b[90m", cyan: "\x1b[96m", blue: "\x1b[95m", green: "\x1b[92m", yellow: "\x1b[93m", red: "\x1b[91m", magenta: "\x1b[95m", orange: "\x1b[96m" },
        "nord" => Ansi { reset: "\x1b[0m", gray: "\x1b[37m", cyan: "\x1b[96m", blue: "\x1b[94m", green: "\x1b[92m", yellow: "\x1b[93m", red: "\x1b[91m", magenta: "\x1b[95m", orange: "\x1b[96m" },
        "mono" => Ansi { reset: "\x1b[0m", gray: "\x1b[90m", cyan: "\x1b[37m", blue: "\x1b[37m", green: "\x1b[37m", yellow: "\x1b[97m", red: "\x1b[97m", magenta: "\x1b[97m", orange: "\x1b[37m" },
        _ => Ansi { reset: "\x1b[0m", gray: "\x1b[90m", cyan: "\x1b[96m", blue: "\x1b[94m", green: "\x1b[92m", yellow: "\x1b[93m", red: "\x1b[91m", magenta: "\x1b[95m", orange: orange_code() },
    }
}

fn color(value: &str, code: &str, ansi: &Ansi) -> String {
    if std::env::var("NO_COLOR").is_ok() || value.is_empty() {
        value.to_string()
    } else {
        format!("{code}{value}{}", ansi.reset)
    }
}

fn level<'a>(value: Option<f64>, ansi: &'a Ansi) -> &'a str {
    match value {
        None => ansi.gray,
        Some(n) if n >= 80.0 => ansi.red,
        Some(n) if n >= 50.0 => ansi.yellow,
        _ => ansi.green,
    }
}

fn tz() -> chrono_tz::Tz {
    std::env::var("CLAUDE_COCKPIT_TZ").ok().and_then(|s| s.parse().ok()).unwrap_or(chrono_tz::Asia::Bangkok)
}

/// with_date: the 7-day reset can be days out, so a bare time is ambiguous —
/// prefix the date. The 5-hour reset is always same-day, so it skips this.
fn reset_clock(epoch_secs: Option<f64>, with_date: bool) -> String {
    let Some(secs) = epoch_secs else { return String::new() };
    let dt = tz().timestamp_opt(secs as i64, 0).single();
    let Some(dt) = dt else { return String::new() };
    if with_date {
        // en-GB's Intl.DateTimeFormat (what the JS version uses) abbreviates
        // September as "Sept", unlike every other month's standard 3-letter
        // form — matched here so the two versions render identically.
        let month = if dt.format("%m").to_string() == "09" { "Sept".to_string() } else { dt.format("%b").to_string() };
        format!("{} {month} {}", dt.format("%e").to_string().trim(), dt.format("%H:%M"))
    } else {
        dt.format("%H:%M").to_string()
    }
}

fn format_duration_ms(value: Option<f64>) -> String {
    let Some(n) = value.filter(|&n| n >= 0.0) else { return String::new() };
    let total_minutes = (n / 60_000.0).floor() as i64;
    format!("{:02}:{:02}", total_minutes / 60, total_minutes % 60)
}

fn vis_len(s: &str) -> usize {
    let mut len = 0;
    let mut chars = s.chars();
    while let Some(c) = chars.next() {
        if c == '\u{1b}' {
            for c2 in chars.by_ref() {
                if c2 == 'm' {
                    break;
                }
            }
        } else {
            len += 1;
        }
    }
    len
}

fn pad_to(s: &str, w: usize) -> String {
    let len = vis_len(s);
    if len >= w { s.to_string() } else { format!("{s}{}", " ".repeat(w - len)) }
}

/// A long session name or branch shouldn't blow out every column's width —
/// cap it before it enters the alignment math.
fn truncate(s: &str, max: usize) -> String {
    let chars: Vec<char> = s.chars().collect();
    if chars.len() > max {
        let head: String = chars[..max - 1].iter().collect();
        format!("{head}…")
    } else {
        s.to_string()
    }
}

fn get_f64(v: &Value, keys: &[&str]) -> Option<f64> {
    let mut cur = v;
    for k in keys {
        cur = cur.get(k)?;
    }
    cur.as_f64().or_else(|| cur.as_str().and_then(|s| s.parse().ok()))
}

fn get_str<'a>(v: &'a Value, keys: &[&str]) -> Option<&'a str> {
    let mut cur = v;
    for k in keys {
        cur = cur.get(k)?;
    }
    cur.as_str()
}

fn first_object<'a>(v: &'a Value, keys: &[&str]) -> Value {
    for k in keys {
        if let Some(obj) = v.get(k) {
            if !obj.is_null() {
                return obj.clone();
            }
        }
    }
    Value::Null
}

pub fn render(input: &Value) -> String {
    let ansi = theme();
    let cwd = get_str(input, &["workspace", "current_dir"])
        .or_else(|| get_str(input, &["cwd"]))
        .map(String::from)
        .unwrap_or_else(|| std::env::current_dir().map(|p| p.to_string_lossy().to_string()).unwrap_or_default());
    let git_info = git_stats(&cwd);
    let model = get_str(input, &["model", "display_name"])
        .or_else(|| get_str(input, &["model", "name"]))
        .or_else(|| input.get("model").and_then(|v| v.as_str()))
        .unwrap_or("--")
        .to_string();
    let context = first_object(input, &["context_window", "context"]);
    let cost = input.get("cost").cloned().unwrap_or(Value::Null);
    let rate = first_object(input, &["rate_limits", "rateLimits"]);
    let five = first_object(&rate, &["five_hour", "5h", "fiveHour"]);
    let seven = first_object(&rate, &["seven_day", "7d", "sevenDay"]);
    let tokens = number(&[
        get_f64(input, &["total_tokens"]),
        get_f64(input, &["tokens"]),
        get_f64(&cost, &["total_tokens"]),
        {
            let a = get_f64(&context, &["total_input_tokens"]);
            let b = get_f64(&context, &["total_output_tokens"]);
            match (a, b) { (Some(a), Some(b)) => Some(a + b), _ => None }
        },
        {
            let a = get_f64(input, &["usage", "input_tokens"]);
            let b = get_f64(input, &["usage", "output_tokens"]);
            match (a, b) { (Some(a), Some(b)) => Some(a + b), _ => None }
        },
    ]);
    let session_start = get_str(input, &["session_start_time"])
        .or_else(|| get_str(input, &["session_start_timestamp"]))
        .or_else(|| get_str(input, &["sessionStartTime"]));
    let clock = {
        let now = tz().from_utc_datetime(&chrono::Utc::now().naive_utc());
        let offset_secs = now.offset().fix().local_minus_utc();
        let sign = if offset_secs < 0 { '-' } else { '+' };
        let h = offset_secs.abs() / 3600;
        let m = (offset_secs.abs() % 3600) / 60;
        let gmt = if m == 0 { format!("GMT{sign}{h}") } else { format!("GMT{sign}{h}:{m:02}") };
        format!("{} {gmt}", now.format("%H:%M"))
    };
    let elapsed = {
        let d = format_duration_ms(get_f64(&cost, &["total_duration_ms"]));
        if !d.is_empty() { d } else { duration_since(session_start) }
    };
    let sep = color(" · ", ansi.gray, &ansi);
    let added = number(&[get_f64(&cost, &["total_lines_added"]), Some(git_info.added as f64)]).unwrap_or(0.0) as u32;
    let removed = number(&[get_f64(&cost, &["total_lines_removed"]), Some(git_info.removed as f64)]).unwrap_or(0.0) as u32;
    let five_reset = reset_clock(number(&[get_f64(&five, &["resets_at"]), get_f64(&five, &["reset_at"])]), false);
    let seven_reset = reset_clock(number(&[get_f64(&seven, &["resets_at"]), get_f64(&seven, &["reset_at"])]), true);
    let context_pct = number(&[get_f64(&context, &["used_percentage"]), get_f64(&context, &["usedPercent"]), get_f64(&context, &["percentage"])]);
    let five_pct = number(&[get_f64(&five, &["used_percentage"]), get_f64(&five, &["usedPercent"]), get_f64(&five, &["percentage"])]);
    let seven_pct = number(&[get_f64(&seven, &["used_percentage"]), get_f64(&seven, &["usedPercent"]), get_f64(&seven, &["percentage"])]);
    let cost_val = number(&[get_f64(&cost, &["total_cost_usd"]), get_f64(&cost, &["totalCostUsd"])]).unwrap_or(0.0);
    let cost_text = format!("${cost_val:.2}");
    let meter = |label: &str, value: Option<f64>, label_color: &str| format!("{} {}", color(label, label_color, &ansi), color(&pct(value), level(value, &ansi), &ansi));

    let session_name = get_str(input, &["session_name"]);
    let project_field = match session_name {
        Some(name) => format!("{}{}", color(&git_info.project, ansi.orange, &ansi), color(&format!("@{name}"), ansi.magenta, &ansi)),
        None => color(&git_info.project, ansi.orange, &ansi),
    };
    let branch_field = if !git_info.branch.is_empty() {
        let changes = if git_info.changes > 0 { format!(" {}", color(&format!("({}±)", git_info.changes), ansi.yellow, &ansi)) } else { String::new() };
        format!("{}{changes}", color(&truncate(&git_info.branch, 16), ansi.orange, &ansi))
    } else {
        color("--", ansi.gray, &ansi)
    };
    let cols1 = [
        branch_field,
        format!("{} {}", color(&format!("+{added}"), ansi.green, &ansi), color(&format!("-{removed}"), ansi.red, &ansi)),
        color(&truncate(&model, 16), ansi.orange, &ansi),
        color(&format!("{clock}{}", if elapsed.is_empty() { String::new() } else { format!(" {elapsed}") }), ansi.blue, &ansi),
        project_field,
    ];
    let cols2 = [
        meter("ctx", context_pct, ansi.cyan),
        color(&compact_tokens(tokens), ansi.cyan, &ansi),
        format!("{}{}", meter("5h", five_pct, ansi.blue), if five_reset.is_empty() { String::new() } else { format!(" {}", color(&five_reset, ansi.blue, &ansi)) }),
        format!("{}{}", meter("7d", seven_pct, ansi.magenta), if seven_reset.is_empty() { String::new() } else { format!(" {}", color(&seven_reset, ansi.magenta, &ansi)) }),
        color(&cost_text, ansi.green, &ansi),
    ];
    // Cap each column so one unusually long field can't stretch every other
    // column with it. The two rows have different field counts on purpose;
    // widths just covers whichever is longer at each index.
    let n = cols1.len().max(cols2.len());
    let widths: Vec<usize> = (0..n)
        .map(|i| vis_len(cols1.get(i).map(String::as_str).unwrap_or("")).max(vis_len(cols2.get(i).map(String::as_str).unwrap_or(""))).min(32))
        .collect();
    let align = |cols: &[String]| -> String {
        let last = cols.len() - 1;
        cols.iter().enumerate().map(|(i, c)| if i == last { c.clone() } else { pad_to(c, widths[i]) }).collect::<Vec<_>>().join(&sep)
    };
    format!("{}\n{}", align(&cols1), align(&cols2))
}

/// Pulls just the two rate-limit windows out of the payload, for the quota
/// cache — kept separate from `render`'s own parsing so the display logic
/// above stays untangled from the caching side effect below.
fn extract_quota(input: &Value) -> crate::quota::Quota {
    let rate = first_object(input, &["rate_limits", "rateLimits"]);
    let five = first_object(&rate, &["five_hour", "5h", "fiveHour"]);
    let seven = first_object(&rate, &["seven_day", "7d", "sevenDay"]);
    crate::quota::Quota {
        five_pct: number(&[get_f64(&five, &["used_percentage"]), get_f64(&five, &["usedPercent"]), get_f64(&five, &["percentage"])]),
        seven_pct: number(&[get_f64(&seven, &["used_percentage"]), get_f64(&seven, &["usedPercent"]), get_f64(&seven, &["percentage"])]),
        five_resets_at: number(&[get_f64(&five, &["resets_at"]), get_f64(&five, &["reset_at"])]),
        seven_resets_at: number(&[get_f64(&seven, &["resets_at"]), get_f64(&seven, &["reset_at"])]),
        updated_ms: now_ms(),
    }
}

pub fn main() {
    let mut raw = String::new();
    use std::io::Read;
    let _ = std::io::stdin().read_to_string(&mut raw);
    let input: Value = if raw.trim().is_empty() { Value::Null } else { serde_json::from_str(&raw).unwrap_or(Value::Null) };
    if let Ok(dir) = std::env::var("CLAUDE_CONFIG_DIR") {
        crate::quota::save(std::path::Path::new(&dir), &extract_quota(&input));
    }
    print!("{}", render(&input));
}
